docs/decisions/0002-auto-enrollment-post-request-approval.rst

```diff
@@ -0,0 +1,113 @@
+
+Automatically enrolling a learner after a request approval ensures that a subsidy is used to enroll a learner in
+the intended course, saves clicks, and provides a more seamless experience.
+
+Decision (?)
+============
+
+We will be adopting event-based communication between the enterprise-access service, the LMS, and the services
+responsible for granting/redeeming a subsidy.
+
+The following events will be sent to an event bus when the state of a subsidy request changes:
+* `org.openedx.enterprise.access.coupon-code-request.pending.v1`
+* `org.openedx.enterprise.access.coupon-code-request.approved.v1`
+* `org.openedx.enterprise.access.license-request.pending.v1``
+* `org.openedx.enterprise.access.license-request.approved.v1``
+These events will be produced via Django signals that live in the `openedx-events <https://github.com/openedx/openedx-events>`__ repo.
+
+Auto-enrollment will be a feature that could be toggled off. It does not always make sense to auto-enroll a learner,
+i.e. when a coupon code offers 50% discount, a license request without course id, etc.. We will add a field `attempt_auto_enrollment`
+to the `SubsidyRequestCustomerConfiguration` model to allow for the auto-enrollment to be turned on/off.
+
+As for handling auto enrollments, there are a few options:
+
+Proposed Solution 1
+===================
+The subsidy granting services will consume the approval events, redeem the granted subsidy,
+and enroll a learner in the course specified in the request.
+
+Changes required
+----------------
+* We will no longer send an email to the learner at the time of approval. Instead, we will send an email to
+  the learner when they have been enrolled.
+* enterprise-access will consume enrollment events from the LMS, check if the enrollment corresponds to an approved
+  request by matching on `lms_user_id`` and `course_id`, and send an approval email to the learner if a matching
+  request is found.
+* We migth not need the above changes if we make the messaging in the approval email generic enough,
+  i.e. "You've been approved, check out the course here!" Learners won't know if we tried to enroll them.
+  Might run into a weird case if a learner tries to redeem a subsidy while we're attempting to auto enroll them.
+* We could add a new state `fulfilled` on the subsidy request models, but unless we need to store that information,
+  I'd propose not changing any of the current models.
+
+Consequences
+------------
+* The learner could see a granted subsidy and enroll manually in course before auto-enrollment occurs.
+* Auto-enrollment could fail and enterprise-access won't receive an enrollment event. The learner will still have their
+  subsidy and can enroll manually but they might not receive an request approved email.
+* Matching an enrollment to a subsidy request based on lms_user_id and course_id is not guaranteed to be reliable.
+  We might want an identifier on the events produced/consumed to tie them to a single interaction.
```

Comment:
My concern with passing an event-specific identifier across service boundaries is that it starts to connect events to one another when, to be truly decoupled, events should be as dumb as possible, and just snapshots of the state of data/thing happening at a given point in time. I also have additional questions of "How would a service know when an event started with itself? Do we store that some place in the service?"

I think the antidote to adding event-tracking complexity comes in the form of generating events with rich enough data that a consuming service can make good decisions around what to do when it sees the events. Something that came up in our chats before was maybe using the source of an enrollment as a way for enterprise-access to understand the nature of the enrollment's creation (e.g. https://github.com/openedx/edx-enterprise/blob/master/enterprise/models.py#L1432)

As a side note, I don't think an event-specific UUID is quite the same as duplicating data across services like we do with enterprise_customer_uuids for example, because `enterprise_customer_uuid` itself is a piece of metadata that has meaning outside of the event lifecycle.

Reply:
I do not believe the identifier has to be an event-specific UUID, but perhaps an identifier that allows us to trace event propagation. I don't think the event becomes less dumb in the presence of something like this. The additional questions are something I have to think about a little more. 

Or perhaps is the problem here that we have a loop in our architecture and perhaps that should be avoided?

Reply:
Agreed that technically adding a UUID doesn't make the event itself less-dumb. And I don't think the looping itself is bad. To me, the crux is more about how/why the looping occurs. 

In the case where we feel the need to add an identifier to trace events across services, I think that means our events are not descriptive enough with the data we have to understand the flow, or we need to categorize our event streams differently. In theory, with just an lms_user_id, enterprise_customer_uuid, and the timestamp for when something is approved in enterprise-access, we should be able to examine logs to understand the flow of what events were produced/consumed across systems (provided we're logging enough)

I worry that an additional identifier bleeds info specific to enterprise-access downstream beyond one consuming service. For example, ecommerce consumes from enterprise-access, so any UUIDs, etc. enterprise-access has is perfectly fine to be expected when ecommerce consumes that event. However, when LMS consumes data from ecommerce, why should the LMS know that the event emitted from ecommerce was triggered by a change in enterprise-access? To truly decouple the services, the LMS can't know what happened before ecommerce emitted its event. Additionally, there is the question again of "what should it do with that tracking UUID in its own data model?"

To me, a tidier solution might be to have a unique enterprise-fulfillment stream (or something like that) that makes the context clear to the LMS the type of event it is consuming without needing to pass extra data between services

Sort of thinking this through as I go. What do you think?

Reply:
We took the conversation offline, updating adr to reflect the conversation.